File: examples/drmaa/drmaa_python/howto4.py
# --------------------------------------------------------------------------
# Copyright 2002-2006 GridWay Team, Distributed Systems Architecture
# Group, Universidad Complutense de Madrid
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# --------------------------------------------------------------------------

from DRMAA import *
import os, sys, time

def setup_job_template():

	args=["-l", "-a"]

	(result, jt, error)=drmaa_allocate_job_template()

	if result != DRMAA_ERRNO_SUCCESS:
		print >> sys.stderr, "drmaa_allocate_job_template() failed: %s" % (error)
		sys.exit(-1)

	cwd=os.getcwd()

	# SHOULD CHECK result, REMOVED FOR THE SAKE OF CLARITY

	(result, error)=drmaa_set_attribute(jt, DRMAA_WD, cwd)
	(result, error)=drmaa_set_attribute(jt, DRMAA_JOB_NAME, "ht4")
	(result, error)=drmaa_set_attribute(jt, DRMAA_REMOTE_COMMAND, "/bin/ls")
	(result, error)=drmaa_set_vector_attribute(jt, DRMAA_V_ARGV, args)

	return jt


########
# MAIN #
########

# ------ INIT DRMAA Library ------
(result, error)=drmaa_init(None)

if result != DRMAA_ERRNO_SUCCESS:
	print >> sys.stderr, "drmaa_init() dailed: #{error}" % (error)
	sys.exit(-1)

jt=setup_job_template()

(result, error)=drmaa_set_attribute(jt, DRMAA_OUTPUT_PATH, 
				"stdout."+DRMAA_PLACEHOLDER_INCR)

(result, error)=drmaa_set_attribute(jt, DRMAA_OUTPUT_PATH, 
				"stdout."+DRMAA_PLACEHOLDER_INCR)


# ------ Submit 5 jobs ------
(result, jobids, error)=drmaa_run_bulk_jobs(jt, 0, 1, 1)

if result != DRMAA_ERRNO_SUCCESS:
	print >> sys.stderr, "drmaa_run_bulk() failed: %s" % (error)
	sys.exit(-1)

print >> sys.stderr, "Bulk job succesfully submited IDs are"

(result, value)=drmaa_get_next_job_id(jobids)
while result != DRMAA_ERRNO_NO_MORE_ELEMENTS:
	if result == DRMAA_ERRNO_SUCCESS:
		print >> sys.stderr, "\t%s" % (value)
		
	(result, value)=drmaa_get_next_job_id(jobids)

print >> sys.stderr, "Waiting for bulk job to finish..."

(result, error)=drmaa_synchronize([DRMAA_JOB_IDS_SESSION_ALL], DRMAA_TIMEOUT_WAIT_FOREVER, 0)

print >> sys.stderr, "All Jobs finished"

(resultj, value)=drmaa_get_next_job_id(jobids)
while resultj != DRMAA_ERRNO_NO_MORE_ELEMENTS:
	if(resultj == DRMAA_ERRNO_SUCCESS):
		(result, job_id_out, stat, rusage, error)=drmaa_wait(value, DRMAA_TIMEOUT_WAIT_FOREVER)
		(result, stat, error)=drmaa_wexitstatus(stat)
		
		print >> sys.stderr, "Rusage for task %s (exit code %s)" % (value, stat)
		
		(result, value)=drmaa_get_next_attr_value(rusage)
		while result != DRMAA_ERRNO_NO_MORE_ELEMENTS:
			if result == DRMAA_ERRNO_SUCCESS:
				print >> sys.stderr, "\t%s" % (value)
				
			(result, value)=drmaa_get_next_attr_value(rusage)

		drmaa_release_attr_values(rusage)
		(resultj, value)=drmaa_get_next_job_id(jobids)
	
drmaa_release_job_ids(jobids)

# ------ Finalize ------
(result, error)=drmaa_delete_job_template(jt)

(result, error)=drmaa_exit()

if result != DRMAA_ERRNO_SUCCESS:
	print >> sys.stderr, "drmaa_exit() failed: %s" % (error)
	sys.exit(-1)

File: examples/drmaa/drmaa_java1.0.1/Howto4.java
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2010, GridWay Project Leads (GridWay.org)                   */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */

import org.ggf.drmaa.*;

import java.util.*;

public class Howto4
{
	public static void main (String[] args)
	{
		SessionFactory	factory = SessionFactory.getFactory();
		Session		session = factory.getSession();
		
		try
		{
			session.init(null);
			System.out.println("Session Init success");
					
			
			JobTemplate jt = session.createJobTemplate();
			
			jt.setWorkingDirectory(java.lang.System.getProperty("user.dir"));
			
			String	    name;		
			jt.setJobName("ht4");
									
			jt.setRemoteCommand("/bin/ls");

			ArrayList   jobArgs = new ArrayList();

			jobArgs.add("-l");
			jobArgs.add("-a");

			jt.setArgs(jobArgs);
			
			jt.setOutputPath("stdout." + GridWaySession.GW_JOB_ID);
			jt.setErrorPath("stderr." + GridWaySession.GW_JOB_ID);
			
			int start = 0;
			int end   = 3;
			int step  = 1;
			int i;
			
			String id;
			
			java.util.List 		ids = session.runBulkJobs(jt, start, end, step);
			java.util.Iterator	iter = ids.iterator();
			
			System.out.println("Bulk job successfully submitted IDs are: ");
			
			while(iter.hasNext())
			{
				System.out.println("\t" + iter.next());
			}
			
			session.deleteJobTemplate(jt);
			
			session.synchronize(Collections.singletonList(Session.JOB_IDS_SESSION_ALL), Session.TIMEOUT_WAIT_FOREVER, false);
			
			for (int count = start; count <= end;count += step)
			{
				JobInfo info = session.wait(Session.JOB_IDS_SESSION_ANY, Session.TIMEOUT_WAIT_FOREVER);
				
				if (info.wasAborted())
					System.out.println("Job " + info.getJobId() + " never ran");
				else if (info.hasExited())
					System.out.println("Job " + info.getJobId() + " finished regularly with exit status " + info.getExitStatus());
				else if (info.hasSignaled())
					System.out.println("Job " + info.getJobId() + " finished due to signal " + info.getTerminatingSignal());
				else 
					System.out.println("Job " + info.getJobId() + " finished with unclear conditions");

				
				System.out.println("Job usage:");
				Map rmap = info.getResourceUsage();
				Iterator r = rmap.keySet().iterator();
				
				while(r.hasNext())
				{
					String name2 = (String) r.next();
					String value = (String) rmap.get(name2);
					System.out.println(" " + name2 + "=" + value);
				}
			}
			session.exit();
			System.out.println("Session Exit success");

		}
		catch (DrmaaException e)
		{
			System.out.println("Error:   " + e.getMessage());
		}
	}
}
